Added onShouldRetry callback for controlling retry

PVince81 opened the pull request:
Whenever the library is about to retry an upload due to an
error, the new optional callback onShouldRetry will be
called when defined. Its return value will tell the library whether
to actually retry the upload or fail with an error, for example
based on status code checks.

This makes it possible to customize the behavior like reacting on
specific status codes.

Fixes https://github.com/tus/tus-js-client/issues/197

PVince81 commented:
the tests are green, seems Travis didn't report back to Github

Acconut commented on README.md:
```diff
@@ -39,6 +39,14 @@ input.addEventListener("change", function(e) {
         onSuccess: function() {
             console.log("Download %s from %s", upload.file.name, upload.url)
+        },
+        onShouldRetry: function(err, retryAttempt, options) {
```
Please move this example to its own section in docs/usage.md. The example in the README.md is meant as the  minimum useful code, so we try to avoid adding too many options here.

Acconut commented on lib/upload.js:
```diff
@@ -923,15 +924,24 @@ function shouldRetry(err, retryAttempt, options) {
   // - we didn't exceed the maxium number of retries, yet, and
   // - this error was caused by a request or it's response and
-  // - the error is server error (i.e. no a status 4xx or a 409 or 423) and
+  // - the error is server error (i.e. no a status 4xx or a 409 or 423) or
```
Can you please help improve this sentence? I think it should be "not a 4XX status except a 409 or 423"

Acconut commented on lib/upload.js:
```diff
@@ -923,15 +924,24 @@ function shouldRetry(err, retryAttempt, options) {
          err.originalRequest != null &&
-         isServerError &&
-         isOnline();
+         isOnline() &&
```
If we already expose a generic function for determining retries, I think we should also remove the online check if the user supplied onShouldRetry. There are use cases where people want to overwrite that check, I believe. What do you think?

Acconut commented on lib/upload.js:
```diff
@@ -923,15 +924,24 @@ function shouldRetry(err, retryAttempt, options) {
-  let isServerError = !inStatusCategory(status, 400) || status === 409 || status === 423;
+
+  // define as function so we only call it if really needed
+  let doRetry = () => {
```
Can you please refactor the code to not use an anonymous function here? I have something like this in mind:

```js
if (retryDelays == null || retryAttempt < retryDelays.lenght || originalRequest == null) {
  return false;
}

if (options && typeof options.onShouldRetry === "function") {
  return options.onShouldRetry(err, retryAttempt, options);
}

return !inStatusCategory(status, 400) || status === 409 || status === 423 && online();
```

That's easier to read IMO.

Acconut commented on test/spec/test-common.js:
```diff
@@ -837,6 +837,130 @@ describe("tus", function () {
+        endpoint: "http://tus.io/files/",
+        retryDelays: [10, 10, 10],
+        onSuccess: waitableFunction("onSuccess"),
+        onShouldRetry: () => true
```
Can you add assertions that onShouldRetry is called with the correct parameters? You can do that using the spyOn functions.

PVince81 commented:
I've addressed all comments.

Had some struggles with `spyOn()` as `callThrough()` didn't work when spying on `upload._emitError` directly to get the error object. It seems that spying on the prototype object worked.

PVince81 commented:
@Acconut anything left to do or is this good to merge ? thanks

PVince81 commented:
@Acconut any concerns merging this addition ?
